components/salesforce/common.js

```diff
@@ -0,0 +1,151 @@
+const slugify = require("slugify");
+const { v4: uuidv4 } = require("uuid");
+
+const salesforce = require("./salesforce.app");
+
+module.exports = {
+  dedupe: "unique",
+  props: {
+    db: "$.service.db",
+    http: {
+      type: "$.interface.http",
+      customResponse: true,
+    },
+    salesforce,
+  },
+  hooks: {
+    async activate() {
+      // Retrieve base API URL's for this account
+      const apiUrls = await this.salesforce.getApiUrls();
+      this.db.set("apiUrls", apiUrls);
+
+      // Enable this Salesforce account to make HTTP
+      // calls to this event source's endpoint.
+      // API docs: https://sforce.co/3jAeV0G
+      const { metadata: metadataApiUrl } = apiUrls;
+      const remoteSiteName = this._getRemoteSiteName();
+      const { endpoint: endpointUrl } = this.http;
+      await this.salesforce.createRemoteSite(metadataApiUrl, remoteSiteName, endpointUrl);
+      this.db.set("remoteSiteName", remoteSiteName);
+
+      // Create a custom Apex class to handle the HTTP
+      // request logic of the webhook.
+      const secretToken = uuidv4();
+      const webhookName = this._getWebhookName();
+      const webhookBody = this.getWebhookBody(webhookName, secretToken);
+      const { id: webhookId } = await this.salesforce.createApexClass(webhookName, webhookBody);
+      this.db.set("secretToken", secretToken);
+      this.db.set("webhookId", webhookId);
+
+      // Create a custom Apex trigger to listen to
+      // the relevant events and dispatch the webhook
+      // call(s) accordingly.
+      const triggerName = this._getTriggerName();
+      const triggerBody = this.getTriggerBody(triggerName, webhookName);
+      const objectType = this.getObjectType();
+      const { id: triggerId } =
+        await this.salesforce.createApexTrigger(triggerName, triggerBody, objectType);
+      this.db.set("triggerId", triggerId);
+    },
+    async deactivate() {
+      const triggerId = this.db.get("triggerId");
+      await this.salesforce.deleteApexTrigger(triggerId);
+
+      const webhookId = this.db.get("webhookId");
+      await this.salesforce.deleteApexClass(webhookId);
+
+      const { metadata: metadataApiUrl } = this.db.get("apiUrls");
+      const remoteSiteName = this.db.get("remoteSiteName");
+      await this.salesforce.deleteRemoteSite(metadataApiUrl, remoteSiteName);
+    },
+  },
+  methods: {
+    _getRemoteSiteName() {
+      const eventSourceName = this.getEventSourceName();
```

Comment:
Won't this throw as written?

Reply:
Oh I see you're overriding in the actual component below

Reply:
Yeah, it's a dirty version of classes with virtual methods :joy: